File: diameterOfABinaryTree.js
// https://leetcode.com/problems/diameter-of-binary-tree/

const diameterOfBinaryTree = function(root) {
	if (!root) return 0;
	const helper = function(node) {
		let left = node.left ? helper(node.left) : 0;
		let right = node.right ? helper(node.right) : 0;
		if (left + right > longest) longest = left + right;
		return left > right ? left + 1 : right + 1;
	}
	let longest = 0;
	helper(root);
	return longest;
}

const TreeNode = function(val) {
	this.val = val;
	this.left = null;
	this.right = null;
}

const test = new TreeNode(1);
const two = new TreeNode(2);
const three = new TreeNode(3);
const four = new TreeNode(4);
const five = new TreeNode(5);
console.assert(diameterOfBinaryTree(test) === 0, "works for a single node");
test.left = two;
test.right = three;
console.assert(diameterOfBinaryTree(test) === 2, "works for a tree of depth two");
two.left = four;
two.right = five;
console.assert(diameterOfBinaryTree(test) === 3, "works");

File: diameterOfABinaryTree.py
# https://leetcode.com/problems/diameter-of-binary-tree/

class TreeNode:
	def __init__(self, val = 0, left = None, right = None):
		self.val = val
		self.left = left
		self.right = right

class Solution:
	def diameterOfBinaryTree(self, root: TreeNode) -> int:
		self.res = 1
		self.recurse(root)
		return self.res - 1

	def recurse(self, node):
		if not node: return 0
		left = self.recurse(node.left)
		right = self.recurse(node.right)
		self.res = max(self.res, left + right + 1)
		return 1 + max(left, right)

test = Solution()
print(test.diameterOfBinaryTree(TreeNode(1, TreeNode(2, TreeNode(4), TreeNode(5)), TreeNode(3))))
